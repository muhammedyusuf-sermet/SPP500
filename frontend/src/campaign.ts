export interface ICampaignState {
	Id?: string,
	Name: string,
	Description: string;
	// Monsters: Monster.IMonsterState[];
}
