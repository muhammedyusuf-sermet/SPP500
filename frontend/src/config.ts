export const API_URL = "https://http://3.18.65.138:3000";