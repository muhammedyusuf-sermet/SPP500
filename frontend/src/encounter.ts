import { IMonsterData } from "./monster";

export interface IEncounterData {
	Id?: string,
	Name: string,
	Description: string;
	Monsters: IMonsterData[];
}
