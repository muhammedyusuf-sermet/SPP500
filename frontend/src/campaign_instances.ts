import * as Campaign from "./campaign";

// Dummy array of monsters
import EncounterInstances from './encounter_instances';

let encounterExample: Campaign.ICampaignState;
encounterExample = {
	Id: 0,
	Name: "Campaign",
	Summary: "Bloody hell...",
	Notes: "You gonna die!",
	Encounters: EncounterInstances['EncounterInstances'],
};

let CampaignInstances = new Array();

// Generate a dummy array of encounters with different IDs
for (var i = 0; i < 12; i++) {
	let encounter = {} as Campaign.ICampaignState;
	Object.assign(encounter, encounterExample);
	encounter['Id'] = i;
	CampaignInstances.push(encounter);
}

export default {
    CampaignInstances,
}