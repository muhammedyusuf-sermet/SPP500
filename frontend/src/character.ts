import { ICampaignData } from "./campaign";

export enum CharacterRace {
	Dragonborn = "Dragonborn",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Gnome = "Gnome",
	HalfElf = "HalfElf",
	Halfling = "Halfling",
	HalfOrc = "HalfOrc",
	Human = "Human",
	Tiefling = "Tiefling"
}

export enum CharacterClass {
	Barbarian = "Barbarian",
	Bard = "Bard",
	Cleric = "Cleric",
	Druid = "Druid",
	Fighter = "Fighter",
	Monk = "Monk",
	Paladin = "Paladin",
	Ranger = "Ranger",
	Rogue = "Rogue",
	Sorcerer = "Sorcerer",
	Warlock = "Warlock",
	Wizard = "Wizard"
}

export interface ICharacterData {
	Id?: number;
	Name: string;
	Level?: number;
	Race?: CharacterRace;
	Class?: CharacterClass;
	MaxHealth?: number;
	ArmorClass?: number;
	Notes?: string;

	Campaigns?: ICampaignData[];
}
