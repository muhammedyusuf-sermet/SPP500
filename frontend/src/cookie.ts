import { API_URL } from "./config";
import Store from 'electron-store';

export class CookieManager {
	private constructor(){
	}

	private static store = new Store<string>();

	public static UserToken(cookieToCheck:string):string {
		// undefined if user is not logged in.
		var cookieVal = CookieManager.store.get(cookieToCheck);

		if (cookieVal === undefined)
			return cookieVal

		var request = require("request");
		var options = { method: 'GET',
			url: API_URL + '/verify',
			timeout: 2000,
			headers:
			{
				'Cache-Control': 'no-cache',
				'Content-Type': 'application/json',
				'Authorization': cookieVal
			},
			body: {},
			json: true
		};

		request(options, function (error: string, response: string, body: {status: number, message: string}) {
			if (error) {
				return undefined
			}
			else {
				return (body.status === 201) ? "Valid cookie" : undefined // Success
			}
		});
		return CookieManager.store.get(cookieToCheck);
	}

	public static SetStringCookie(name:string, data:string) {
		CookieManager.store.set(name, data);
	}

	public static RemoveCookie(name:string) {
		CookieManager.store.delete(name);
	}
}
