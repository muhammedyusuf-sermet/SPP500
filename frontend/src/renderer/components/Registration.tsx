import * as React from 'react';
import { Title, Modal, ModalBackground, ModalContent, Box, Field, Control, Input, Button, Help } from 'bloomer';

import '../css/registration.css';
import { API_URL } from '../../config';
import 'bulma/css/bulma.css';


export interface IRegisterState {
	user: {
		username: string,
		password: string,
		email: string,
		name: string
	}
	modal: {
		open: boolean,
		message: string
	}
}

interface IRegisterResponce {
	status: number,
	messages: string[]
}

export class Registration extends React.Component<any, IRegisterState> {
	constructor(props: any) {
		super(props);
		this.state = {
			user: {
				username: "",
				password: "",
				email: "",
				name: ""
			},

			modal: {
				open: false,
				message: ""
			}
		}
	}

	handleUsernameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const user = this.state.user
		this.setState({
			user: { ...user, username: event.target.value }
		})
	}

	handlePasswordChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const user = this.state.user
		this.setState({
			user: { ...user, password: event.target.value }
		})
	}

	handleEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const user = this.state.user
		this.setState({
			user: { ...user, email: event.target.value }
		})
	}

	handleNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const user = this.state.user
		this.setState({
			user: { ...user, name: event.target.value }
		})
	}

	/*Todo: Create a new helper component for Snackbars - reuse it*/
	openModal = (messageText: string) => {
		const modal = this.state.modal
		this.setState({ modal: {...modal, open: true, message: messageText }});
	};

	closeModal = () => {
		const modal = this.state.modal
		this.setState({ modal: {...modal, open: false }});
	};

	requestRegister = (event: React.FormEvent) => {
		event.preventDefault();
		this.processRegister();
	}

	processRegister = async () => {

		var request = require('request-promise-native');
		var options = { method: 'POST',
			url: API_URL + '/register',
			headers:
			{
				'Cache-Control': 'no-cache',
				'Content-Type': 'application/json'
			},
			body:
			{
				username: this.state.user.username,
				password: this.state.user.password,
				email: this.state.user.email,
				name: this.state.user.name
			},
			json: true
		};

		await request(options)
				.then((body: IRegisterResponce) => {
					if (body.status == 201) { // success
						this.openModal("Welcome aboard! You can now login with your username and password.");
					} else if (body.messages) {
						this.openModal("Error registering: " + body.messages.toString());
					}else{
						this.openModal("There was an error creating your account. Please try again later.")
					}
				})
				.catch((error: string) => {
					this.openModal("There was an error sending your request. Please try again later.");
					// throw new Error(error);
				});
	}

	render() {
		
		return (
			<React.Fragment>
				<form onSubmit={this.requestRegister}>
					<Title>Register Now!</Title>
					<Field>
						<Control >
							<Input
								id='name'
								type='text'
								placeholder='Name'
								autoComplete='name'
								value={this.state.user.name}
								onChange={this.handleNameChange}
								required />

						</Control>
					</Field>
					<Field>
						<Control>
							<Input
								id='email'
								type='email'
								placeholder='E-mail Address'
								autoComplete='email'
								value={this.state.user.email}
								onChange={this.handleEmailChange}
								required />
								<Help>Emails must be provided in the x@y.z format and must be unique.</Help>
						</Control>
					</Field>
					<Field isGrouped='centered' isHorizontal>
						<Control isExpanded>
							<Input
								id='username'
								type='username'
								placeholder='Username'
								autoComplete='username'
								value={this.state.user.username}
								onChange={this.handleUsernameChange}
								required />
							<Help>Usernames must be unique and have at least 1 character.</Help>
						</Control>
						<Control isExpanded>
							<Input
								id='password'
								type='password'
								placeholder='Password'
								autoComplete='password'
								value={this.state.user.password}
								onChange={this.handlePasswordChange}
								required />
							<Help>Passwords must be over 7 characters long.</Help>
						</Control>
					</Field>
					<Field>
						<Button isColor='primary' type="submit" isLoading={false}>Register</Button>
					</Field>
				</form>
				<Modal id='registerModal' isActive={this.state.modal.open}>
					<ModalBackground id='modalBackground' onClick={()=>{
						this.closeModal();
					}}/>
					<ModalContent>
						<Box>
							<span id="ModalMessage">{this.state.modal.message}</span>
						</Box>
					</ModalContent>
				</Modal>
			</React.Fragment>
		);
	}
}
