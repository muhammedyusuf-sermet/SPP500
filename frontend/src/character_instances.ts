import { CharacterRace, CharacterClass, ICharacterData } from "./character";


let characterExample: ICharacterData;
characterExample = {
	Id: 1,
	Name: "Character",
	Level: 3,
	Race: CharacterRace.Gnome,
	Class: CharacterClass.Paladin,
	MaxHealth: 23,
	ArmorClass: 12,
	Notes: 'aggressive',

	Campaigns: []
};

export const CharacterInstances: ICharacterData[] = new Array();

// Generate a dummy array of characters with different IDs
for (var i = 0; i < 12; i++) {
	let character = {} as ICharacterData;
	Object.assign(character, characterExample);
	character.Id = i + 1;
	CharacterInstances.push(character);
}