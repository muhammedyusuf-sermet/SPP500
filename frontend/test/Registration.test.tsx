import * as React from "react"
import * as nock from 'nock';
import {shallow, ShallowWrapper} from 'enzyme';
import { Registration, IRegisterState } from "../src/renderer/components/Registration";
import { API_URL } from "../src/config";

jest.mock('../src/cookie');

describe('Register Component', () => {
	let registerInstance: ShallowWrapper<any, IRegisterState, Registration>;

	beforeEach(() => {
		registerInstance = shallow(<Registration/>);
	});

	it ('matches snapshot', () => {
		expect(registerInstance).toMatchSnapshot();
	});

	it('renders without crashing', () => {
		expect(registerInstance).toBeDefined();
	});

	describe('should respond to change event and change the state of the Register Component', () => {
		it('all input fields work', () => {
			var usernameBox = registerInstance.find('#username');
			var passwordBox = registerInstance.find('#password');
			var emailBox = registerInstance.find('#email');
			var nameBox = registerInstance.find('#name');

			usernameBox.simulate('change', {target: {name: 'username', value: 'test_username'}});
			passwordBox.simulate('change', {target: {name: 'password', value: 'test_password'}});
			emailBox.simulate('change', {target: {name: 'username', value: 'test_email'}});
			nameBox.simulate('change', {target: {name: 'password', value: 'test_name'}});

			expect(registerInstance.state('user')).toEqual({
				password: 'test_password',
				username: 'test_username',
				email: 'test_email',
				name: 'test_name'
			});
		});
	});

	describe('should show and hide modal', () => {
		it('show modal', () => {
			registerInstance.instance().closeModal();
			expect(registerInstance.find('#registerModal').prop('isActive')).toEqual(false);
			registerInstance.instance().openModal('TestMessage');
			expect(registerInstance.find('#ModalMessage').text()).toEqual('TestMessage');
			expect(registerInstance.find('#registerModal').prop('isActive')).toEqual(true);
		})

		it('close modal', () => {
			registerInstance.instance().openModal('TestMessage');
			expect(registerInstance.find('#ModalMessage').text()).toEqual('TestMessage');
			expect(registerInstance.find('#registerModal').prop('isActive')).toEqual(true);
			registerInstance.instance().closeModal();
			expect(registerInstance.find('#registerModal').prop('isActive')).toEqual(false);
		})

		it('close modal by click', () => {
			registerInstance.instance().openModal('TestMessage');
			expect(registerInstance.find('#ModalMessage').text().length).toBeGreaterThan(0);
			expect(registerInstance.find('#registerModal').prop('isActive')).toEqual(true);
			let background = registerInstance.find('#modalBackground')
			background.simulate('click');
			expect(registerInstance.find('#registerModal').prop('isActive')).toEqual(false);
		})
	});

	describe('makes a server request to register the user', () => {

		beforeEach(() => {
			nock.disableNetConnect();
			//let scope: nock.Scope;
			nock(API_URL)
			.post('/register', {
					username: "test_username",
					password: "test_password",
					email   : "test_email",
					name    : "test_name"
				})
			.reply(201,
				{ status: 201, messages: 'success' },
			);
		});

		afterEach(() => {
			/*
			This works to detect when the nock is not used.
			However the register function is async so getting
				an expect to do anything useful was impossible.
			if(scope)
				expect(scope.pendingMocks()).toEqual([]);
			*/
		})

		it('successfully register with correct credentials', () => {
			var usernameBox = registerInstance.find('#username');
			var passwordBox = registerInstance.find('#password');
			var emailBox = registerInstance.find('#email');
			var nameBox = registerInstance.find('#name');
			var registerForm = registerInstance.find('form');

			usernameBox.simulate('change', {target: {name: 'username', value: 'test_username'}});
			passwordBox.simulate('change', {target: {name: 'password', value: 'test_password'}});
			emailBox.simulate('change', {target: {name: 'username', value: 'test_email'}});
			nameBox.simulate('change', {target: {name: 'password', value: 'test_name'}});

			nock(API_URL)
				.post('/register', {
						"username": "test_username",
						"password": "test_password",
						"email"   : "test_email",
						"name"    : "test_name"
					})
				.reply(201, {
					body: [{ status: 201, messages: 'success' }],
				});
			registerForm.simulate('submit', {preventDefault: () => {}});
		});
	});
});
