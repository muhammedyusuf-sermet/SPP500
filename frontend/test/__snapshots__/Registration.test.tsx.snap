// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component matches snapshot 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <withHelpersModifiers(Title)>
      Register Now!
    </withHelpersModifiers(Title)>
    <withHelpersModifiers(Field)>
      <withHelpersModifiers(Control)>
        <withHelpersModifiers(Input)
          autoComplete="name"
          id="name"
          onChange={[Function]}
          placeholder="Name"
          required={true}
          type="text"
          value=""
        />
      </withHelpersModifiers(Control)>
    </withHelpersModifiers(Field)>
    <withHelpersModifiers(Field)>
      <withHelpersModifiers(Control)>
        <withHelpersModifiers(Input)
          autoComplete="email"
          id="email"
          onChange={[Function]}
          placeholder="E-mail Address"
          required={true}
          type="email"
          value=""
        />
        <withHelpersModifiers(Help)>
          Emails must be provided in the x@y.z format and must be unique.
        </withHelpersModifiers(Help)>
      </withHelpersModifiers(Control)>
    </withHelpersModifiers(Field)>
    <withHelpersModifiers(Field)
      isGrouped="centered"
      isHorizontal={true}
    >
      <withHelpersModifiers(Control)
        isExpanded={true}
      >
        <withHelpersModifiers(Input)
          autoComplete="username"
          id="username"
          onChange={[Function]}
          placeholder="Username"
          required={true}
          type="username"
          value=""
        />
        <withHelpersModifiers(Help)>
          Usernames must be unique and have at least 1 character.
        </withHelpersModifiers(Help)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)
        isExpanded={true}
      >
        <withHelpersModifiers(Input)
          autoComplete="password"
          id="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
          value=""
        />
        <withHelpersModifiers(Help)>
          Passwords must be over 7 characters long.
        </withHelpersModifiers(Help)>
      </withHelpersModifiers(Control)>
    </withHelpersModifiers(Field)>
    <withHelpersModifiers(Field)>
      <withHelpersModifiers(Button)
        isColor="primary"
        isLoading={false}
        type="submit"
      >
        Register
      </withHelpersModifiers(Button)>
    </withHelpersModifiers(Field)>
  </form>
  <withHelpersModifiers(Modal)
    id="registerModal"
    isActive={false}
  >
    <withHelpersModifiers(ModalBackground)
      id="modalBackground"
      onClick={[Function]}
    />
    <withHelpersModifiers(ModalContent)>
      <withHelpersModifiers(Box)>
        <span
          id="ModalMessage"
        />
      </withHelpersModifiers(Box)>
    </withHelpersModifiers(ModalContent)>
  </withHelpersModifiers(Modal)>
</Fragment>
`;
